### ISSUE TITLE
Einsum functionality

### ISSUE THREAD

**Issue by MannyKayy** [cupy/cupy#19]
Any plan to support numpy einsum soon?

I noticed a comment about it in [product.py](https://github.com/cupy/cupy/blob/update-v1.21.0/cupy/linalg/product.py#L196)

Thanks

**Comment by jekbradbury** [comment]
May be able to take inspiration from Harvard `autograd` toolkit, which implements both `np.einsum` and its gradient

**Comment by okuta** [comment]
Sorry, there are no plan.

PR welcome.

**Comment by muupan** [comment]
Tensorflow has einsum. I want it in chainer. https://www.tensorflow.org/api_docs/python/tf/einsum

**Comment by corochann** [comment]
Thank you for the einsum implementation in cupy!
But I found that its memory performance is not optimized compared to another function.

`
import argparse

import cupy

parser = argparse.ArgumentParser(description='Chainer example: Cifar')
parser.add_argument('--batchsize', '-b', type=int, default=1000)
args = parser.parse_args()

batchsize = args.batchsize

gy = cupy.random.randint(-100, 100, (batchsize, 32, 3, 3))
col = cupy.random.randint(-100, 100, (batchsize, 32, 3, 3, 3, 3))
gW_original = cupy.tensordot(
    gy, col, ((0, 2, 3), (0, 4, 5)))

print('calc matmul...')
y0, y1, y2, y3 = gy.shape
c0, c1, c2, c3, c4, c5 = col.shape
gW_matmul = cupy.matmul(gy.reshape(y0, y1, y2*y3), col.reshape(c0, c1 * c2 * c3, -1).transpose(0, 2, 1)).reshape(y0, y1, c1, c2, c3)
print('calc matmul ok')

print('calc einsum...')
gW_einsum = cupy.einsum('ijkl,imnokl->ijmno', gy, col)
print('calc einsum ok')

#print(gW_original.shape, gW_einsum.shape)
print('diff', cupy.sum(gW_einsum, axis=0) - gW_original)
print('diff2', cupy.sum(gW_matmul, axis=0) - gW_original)

#assert cupy.allclose(cupy.sum(gW_einsum, axis=0), gW_original)
#assert cupy.allclose(cupy.sum(gW_matmul, axis=0), gW_original)
`


einsum got cupy.cuda.runtime.CUDARuntimeError: cudaErrorMemoryAllocation: out of memory
with batchsize=100, while matmul did not get memory error even with batchsize = 50000.

Please check with the code,
`python cupytest.py -b 100`

**Comment by fukatani** [comment]
@corochann 
Thank you for your benchmark.
Current einsum implementation is simple and not optimized.

Especially, `calc_combined_view` is slow.
Once calculate outer view with `cupy.tensordot(result, ioperand, axes=0)`, and reduce by `diagnonal` and `sum` if needed.
But it become faster if we use `matmul` directly instead of current three stage approach.

I'll start after #404 and #410 merged.
Of cource PR is welcome.

**Comment by niboshi** [comment]
We have the implementation now, so I'm closing the issue.

**Comment by fukatani** [comment]
@corochann
Now, we optimized `einsum` partially, for broadcasted axes.
I mean that `cupy.einsum('...jkl,...mnokl->...jmno', gy, col)` use less memory than `cupy.einsum('ijkl,imnokl->ijmno', gy, col)`.
And I confirmed to finish with batchsize=500 by GTX 1080 (8GB).

I recommend to use '...' if enable in current cupy.einsum implementation.
Of course, we can optimize `cupy.einsum` more, not only for broadcasted axes.
